### DISCUSSION TITLE
LSP signature help highlights wrong words in name of function.

### DISCUSSION THREAD

**Discussion by nyeong** [helix-editor/helix#10398] [General]
### Summary

<img width="658" alt="스크린샷 2022-10-05 17 02 45" src="https://user-images.githubusercontent.com/12120034/194010751-bdcea589-e253-4c6c-8465-47156d43bf80.png">

After 22.08, *LSP signature help* has come, which highlights the function parameters which is currently being entered.
*Signature help* highlights the words in the function name if there are the same words in the parameters.

### Reproduction Steps

 

It can be reproduced by defining a function named with the words in parameters. I can reproduce it with elixir-lsp, helix 22.08.1. Other LSP should be tested.

https://user-images.githubusercontent.com/12120034/194015259-836612af-bfa8-49ac-8412-ec0a541daa7b.mov

### Helix log

There is no log about this.

### Platform

macOS 12.6 (Apple Silicon)

### Terminal Emulator

iTerm 3.4.16

### Helix Version

helix 22.08.1 (66276ce6)

**Comment by nyeong** [comment]
https://github.com/helix-editor/helix/blob/4a3b776b78c68f21017b1900eafb7fefeadeefa8/helix-term/src/commands/lsp.rs#L891-L912

`signature_help_impl` function is related.

**Comment by A-Walrus** [comment]
Played around with this in VS Code and a similar thing occurs, but only when the name of the argument is the entire name of the function (which may or may not be valid code depending on your language but whatever).
![image](https://user-images.githubusercontent.com/58790821/194269670-2d7c821a-4cbc-4722-8440-e29f875478cb.png)

I believe that the LSP specification is flawed. 
According to the spec [specification#parameterInformation](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#parameterInformation), language servers can pass either the string label of the parameter, or the range of characters for it. If they pass the label as a string the client is expected to find the substring within the full label for the signature help and highlight it. This is problematic as the editor could mistakenly find substrings within the name of the function, within another parameter which contains that parameter name, within keywords such as `fn` `def` `do` `func` ...

We could try to make better guesses as to where and how we search for the substring, but ultimately I think the specification needs to be changed so that passing the parameter label as a string is deprecated and instead all language servers pass the character range.
